### PR TITLE
registry: Refactor requestfactory to use http.RoundTrippers

### PR DESCRIPTION
**Pull request by tiborvass**
Replaces #13254 (messed up github)

This patch removes the need for requestFactories and decorators
by implementing http.RoundTripper transports instead.

It refactors some challenging-to-read code.

NewSession now takes an *http.Client that can already have a
custom Transport, it will add its own auth transport by wrapping
it.

The idea is that callers of http.Client should not bother
setting custom headers for every handler but instead it should
be transparent to the callers of a same context.

This patch is needed for future refactorings of registry,
namely refactoring of the v1 client code.

Signed-off-by: Tibor Vass tibor@docker.com

### CODE REVIEW

File: registry/registry.go [hunk 1]
```diff
@@ -31,65 +35,17 @@ const (
...
-		Transport:     httpTransport,
-		CheckRedirect: AddRequiredHeadersToRedirectedRequests,
-		Jar:           jar,
-	}
+type httpsTransport struct {
+	*http.Transport
 }
 
-func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secure bool) (*http.Response, *http.Client, error) {
+func (tr *httpsTransport) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (tiborvass) [inline registry/registry.go]: If someone wonders why do we set tlsconfig in a roundtrip, it's because it's so as to match the current behavior in master: we generate the certpool on every-goddam-request. It's not great, but it allows people to just put the certs in /etc/docker/certs.d/.../ and let docker "pick it up" immediately. Would prefer an fsnotify implementation, but that's not the scope of this PR

File: graph/pull.go [hunk 2]
```diff
@@ -60,7 +60,8 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
 		return err
 	}
 
-	r, err := registry.NewSession(imagePullConfig.AuthConfig, registry.HTTPRequestFactory(imagePullConfig.MetaHeaders), endpoint, true)
+	client := registry.HTTPClient(&registry.DockerHeaders{registry.DefaultTransport(registry.ReceiveTimeout, endpoint.IsSecure), imagePullConfig.MetaHeaders})
```
Comment (stevvooe) [inline graph/pull.go]: Might want to break this up over a few lines to let it breath (readability) and bring up your SLOC stats. ;)

File: graph/pull.go [hunk 3]
```diff
@@ -240,7 +241,7 @@ func (s *TagStore) pullRepository(r *registry.Session, out io.Writer, repoInfo *
 }
 
 func (s *TagStore) pullImage(r *registry.Session, out io.Writer, imgID, endpoint string, token []string, sf *streamformatter.StreamFormatter) (bool, error) {
-	history, err := r.GetRemoteHistory(imgID, endpoint, token)
```
Comment (stevvooe) [inline graph/pull.go]: Yes! This is great!

File: graph/push.go [hunk 4]
```diff
@@ -514,7 +514,8 @@ func (s *TagStore) Push(localName string, imagePushConfig *ImagePushConfig) erro
 		return err
 	}
 
-	r, err := registry.NewSession(imagePushConfig.AuthConfig, registry.HTTPRequestFactory(imagePushConfig.MetaHeaders), endpoint, false)
+	client := registry.HTTPClient(&registry.DockerHeaders{registry.DefaultTransport(registry.NoTimeout, endpoint.IsSecure), imagePushConfig.MetaHeaders})
```
Comment (stevvooe) [inline graph/push.go]: Break this line up, as well.

File: pkg/requestdecorator/requestdecorator.go [hunk 5]
```diff
@@ -47,7 +47,7 @@ func (vi *UAVersionInfo) isValid() bool {
 // "product/version", where the "product" is get from the name field, while
 // version is get from the version field. Several pieces of verson information
 // will be concatinated and separated by space.
-func appendVersions(base string, versions ...UAVersionInfo) string {
+func AppendVersions(base string, versions ...UAVersionInfo) string {
```
Comment (stevvooe) [inline pkg/requestdecorator/requestdecorator.go]: Make sure to update the Godoc.

File: registry/auth.go [hunk 6]
```diff
@@ -214,7 +212,7 @@ func loginV1(authConfig *cliconfig.AuthConfig, registryEndpoint *Endpoint, facto
 // now, users should create their account through other means like directly from a web page
 // served by the v2 registry service provider. Whether this will be supported in the future
 // is to be determined.
-func loginV2(authConfig *cliconfig.AuthConfig, registryEndpoint *Endpoint, factory *requestdecorator.RequestFactory) (string, error) {
```
Comment (stevvooe) [inline registry/auth.go]: Here is another example of this PR doing its job: `authConfig` and `registryEndpoint` are structurally related to login. The `factory` is is not.

File: registry/auth.go [hunk 7]
```diff
@@ -268,13 +266,13 @@ func tryV2BasicAuthLogin(authConfig *cliconfig.AuthConfig, params map[string]str
 	return nil
 }
 
-func tryV2TokenAuthLogin(authConfig *cliconfig.AuthConfig, params map[string]string, registryEndpoint *Endpoint, client *http.Client, factory *requestdecorator.RequestFactory) error {
-	token, err := getToken(authConfig.Username, authConfig.Password, params, registryEndpoint, client, factory)
+func tryV2TokenAuthLogin(authConfig *cliconfig.AuthConfig, params map[string]string, registryEndpoint *Endpoint, client *http.Client) error {
```
Comment (stevvooe) [inline registry/auth.go]: This may be out of scope for this PR: `authConfig` is the first argument to several functions. Is there a missing struct?

```
type loginController struct {
    authConfig *cliconfig.AuthConfig
}
```

File: registry/endpoint.go [hunk 8]
```diff
@@ -171,12 +168,12 @@ func (e *Endpoint) pingV1(factory *requestdecorator.RequestFactory) (RegistryInf
...
 	}
 
-	req, err := factory.NewRequest("GET", e.Path("_ping"), nil)
+	req, err := http.NewRequest("GET", e.Path("_ping"), nil)
 	if err != nil {
 		return RegistryInfo{Standalone: false}, err
 	}
 
-	resp, _, err := doRequest(req, nil, ConnectTimeout, e.IsSecure)
+	resp, err := HTTPClient(DefaultTransport(ConnectTimeout, e.IsSecure)).Do(req)
```
Comment (stevvooe) [inline registry/endpoint.go]: Should this client have been passed in, rather than instantiated here? Or does it not require the same `RoundTripper`?

File: registry/endpoint.go [hunk 9]
```diff
@@ -265,18 +262,6 @@ HeaderLoop:
...
-	return &http.Client{
-		Transport: &http.Transport{
-			DisableKeepAlives: true,
-			Proxy:             http.ProxyFromEnvironment,
-			TLSClientConfig:   &tlsConfig,
-		},
-		CheckRedirect: AddRequiredHeadersToRedirectedRequests,
-	}
+	// TODO: shouldn't this be ConnectTimeout for safety?
+	return HTTPClient(DefaultTransport(NoTimeout, e.IsSecure))
```
Comment (stevvooe) [inline registry/endpoint.go]: Maybe this:

```
if e.client == nil {
    e.client = HTTPClient(DefaultTransport(NoTimeout, e.IsSecure))
}

return e.client
```

File: registry/endpoint.go [hunk 10]
```diff
@@ -265,18 +262,6 @@ HeaderLoop:
...
-	}
-	return &http.Client{
-		Transport: &http.Transport{
-			DisableKeepAlives: true,
-			Proxy:             http.ProxyFromEnvironment,
-			TLSClientConfig:   &tlsConfig,
-		},
-		CheckRedirect: AddRequiredHeadersToRedirectedRequests,
-	}
+	// TODO: shouldn't this be ConnectTimeout for safety?
```
Comment (stevvooe) [inline registry/endpoint.go]: Can't hurt.

File: registry/registry.go (hunk 1 above)
Comment (stevvooe) [inline registry/registry.go]: Maybe drop that in a comment here:

```
DRAGONS(tiborvass): If someone wonders why do we set tlsconfig in a roundtrip, 
it's because it's so as to match the current behavior in master: we generate the
certpool on every-goddam-request. It's not great, but it allows people to just put 
the certs in /etc/docker/certs.d/.../ and let docker "pick it up" immediately. Would 
prefer an fsnotify implementation, but that's not the scope of this PR.
```

That would minimize the later WTFs and provide some direction in fixing this.

File: registry/registry.go [hunk 11]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
+				MinVersion: tls.VersionTLS10,
+			}
+		}
+		tr.Transport.TLSClientConfig.RootCAs = roots
+		tr.Transport.TLSClientConfig.Certificates = certs
+	}
+	return tr.Transport.RoundTrip(req)
+}
+
+func DefaultTransport(timeout TimeoutType, secure bool) http.RoundTripper {
```
Comment (stevvooe) [inline registry/registry.go]: I would call this `NewTransport` or something like that to match Go conventions. A name like `DefaultTransport` implies a global that is ready to use.

File: registry/registry.go [hunk 12]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
+		transport.Dial = func(proto string, addr string) (net.Conn, error) {
+			// Set the connect timeout to 5 seconds
+			d := net.Dialer{Timeout: 5 * time.Second, DualStack: true}
+
+			conn, err := d.Dial(proto, addr)
+			if err != nil {
+				return nil, err
+			}
+			// Set the recv timeout to 10 seconds
+			conn.SetDeadline(time.Now().Add(10 * time.Second))
```
Comment (stevvooe) [inline registry/registry.go]: I know this was here before, but I am not sure that setting the deadline here is the right call. Shouldn't we rely on `http.Client.Timeout`?  I think this gets set by the `http` package.

File: registry/registry.go [hunk 13]
```diff
@@ -136,24 +92,135 @@ func doRequest(req *http.Request, jar http.CookieJar, timeout TimeoutType, secur
...
+	for k, v := range tr.Headers {
+		req.Header[k] = v
+	}
+	return tr.RoundTripper.RoundTrip(req)
+}
+
+type debugTransport struct{ http.RoundTripper }
+
+func (tr debugTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	fmt.Println(req.Method, req.URL)
```
Comment (stevvooe) [inline registry/registry.go]: There is a function called `httputil.DumpRequest` that can do this for you.

File: graph/pull.go (hunk 2 above)
Comment (tiborvass) [inline graph/pull.go]: ok, lol I just removed most of the requestfactory package so that should help SLOC :P

File: registry/registry.go (hunk 1 above)
Comment (tiborvass) [inline registry/registry.go]: will do thanks

File: registry/auth.go (hunk 6 above)
Comment (tiborvass) [inline registry/auth.go]: i love your pr reviews

File: registry/auth.go (hunk 7 above)
Comment (tiborvass) [inline registry/auth.go]: Agreed, will add a TODO

File: registry/session.go [hunk 14]
```diff
@@ -20,64 +21,96 @@ import (
 	"github.com/Sirupsen/logrus"
 	"github.com/docker/docker/cliconfig"
 	"github.com/docker/docker/pkg/httputils"
-	"github.com/docker/docker/pkg/requestdecorator"
```
Comment (stevvooe) [inline registry/session.go]: Can we delete the entire `requestdecorator` package with this PR?

File: registry/endpoint.go (hunk 8 above)
Comment (tiborvass) [inline registry/endpoint.go]: doRequest instantiates a new client everysingle time which is weird indeed. Initially i made it so that there was only one client, but I'm not 100% sure it's a good idea, either way thought it's out of scope unless something is actually broken.

However, everything in session.go reuses the same client embedded in the Session struct. That is so that the same cookiejar is reused, like it was the case on master by passing the jar around every function. On this line `doRequest` has a `nil` jar, which is what `HTTPClient()` does by default.

I'd be happy to revisit this in another PR. (again: unless there is something fundamentally wrong)

File: registry/endpoint.go (hunk 9 above)
Comment (tiborvass) [inline registry/endpoint.go]: Oh well, I don't see why this would be an issue and I do prefer it. Hopefully that resolves https://github.com/docker/docker/pull/13265#discussion_r30450466

File: registry/registry.go (hunk 11 above)
Comment (tiborvass) [inline registry/registry.go]: ok

File: registry/session.go [hunk 15]
```diff
@@ -20,64 +21,96 @@ import (
...
 )
 
 type Session struct {
-	authConfig    *cliconfig.AuthConfig
-	reqFactory    *requestdecorator.RequestFactory
 	indexEndpoint *Endpoint
-	jar           *cookiejar.Jar
-	timeout       TimeoutType
+	client        *http.Client
+	// TODO: remove authConfig
```
Comment (stevvooe) [inline registry/session.go]: `authConfig` definitely needs a home but I am not sure where it belongs. Seems like either on `Session` or `Service`. Typically, I would layer this as part of the `Session` but I am not sure if that is the right place for this package.

File: registry/registry.go (hunk 12 above)
Comment (tiborvass) [inline registry/registry.go]: I would prefer considering this out of scope of refactoring

File: registry/registry.go (hunk 13 above)
Comment (tiborvass) [inline registry/registry.go]: I <3 you

File: registry/session.go [hunk 16]
```diff
@@ -147,44 +165,43 @@ func (r *Session) GetRemoteImageJSON(imgID, registry string, token []string) ([]
...
 	)
 
-	req, err := r.reqFactory.NewRequest("GET", imageURL, nil)
+	req, err := http.NewRequest("GET", imageURL, nil)
 	if err != nil {
-		return nil, fmt.Errorf("Error while getting from the server: %s\n", err)
+		return nil, fmt.Errorf("Error while getting from the server: %v", err)
 	}
-	setTokenAuth(req, token)
+	// TODO: why are we doing retries at this level?
```
Comment (stevvooe) [inline registry/session.go]: What level should they be at? (Please specify in TODO comment).

File: registry/endpoint.go (hunk 8 above)
Comment (stevvooe) [inline registry/endpoint.go]: That sounds okay: `http.Client` is cheap to instantiate. The expensive piece is the transport, since you may disassociate from the connection pool. Seems like there should be a common `http.Transport` instance to support pooling but I do understand there are issues with the certs are handled.

If this was the previous behavior, no reason to change it for this PR.

File: registry/registry.go (hunk 12 above)
Comment (stevvooe) [inline registry/registry.go]: agreed.

File: registry/auth.go (hunk 7 above)
Comment (tiborvass) [inline registry/auth.go]: actually doesnt matter, it should be refactored in the distrib client vendoring pr

**Comment by stevvooe** [comment]
There a few style nits but this looks ready to go.

LGTM!

File: registry/session.go (hunk 14 above)
Comment (tiborvass) [inline registry/session.go]: see added commit

File: registry/session.go (hunk 15 above)
Comment (tiborvass) [inline registry/session.go]: right, let's do it in another PR :)

**Comment by stevvooe** [comment]
@tiborvass @diogomonica has a few comments in https://github.com/docker/docker/pull/13254#discussion_r30455417 that might be interesting.

File: pkg/useragent/README.md [hunk 17]
```diff
@@ -0,0 +1 @@
+This package provides helper functions to pack version information into a single User-Agent header.
```
Comment (stevvooe) [inline pkg/useragent/README.md]: This should just be in `doc.go`.

**Comment by stevvooe** [comment]
LGTM!

File: registry/session.go (hunk 16 above)
Comment (tiborvass) [inline registry/session.go]: added

File: pkg/useragent/README.md (hunk 17 above)
Comment (tiborvass) [inline pkg/useragent/README.md]: this was a birthday party issue i believe

**Comment by tiborvass** [comment]
The failure seems unrelated, yoohoo!!

**Comment by tiborvass** [comment]
@diogomonica took care of your comments

File: pkg/transport/transport.go [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+type headerModifier http.Header
+
+// NewHeaderRequestModifier returns a RequestModifier that merges the HTTP headers
+// passed as an argument, with the HTTP headers of a request.
+//
+// If the same key is present in both, the modifying header values for that key,
+// are appended to the values for that same key in the request header.
+func NewHeaderRequestModifier(header http.Header) RequestModifier {
```
Comment (tiborvass) [inline pkg/transport/transport.go]: added function comment

File: pkg/transport/transport.go [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+		req.Header[k] = append(req.Header[k], s...)
+	}
+
+	return nil
+}
+
+// NewTransport returns an http.RoundTripper that modifies requests according to
+// the RequestModifiers passed in the arguments, before sending the requests to
+// the base http.RoundTripper (which, if nil, defaults to http.DefaultTransport).
+func NewTransport(base http.RoundTripper, modifiers ...RequestModifier) http.RoundTripper {
```
Comment (tiborvass) [inline pkg/transport/transport.go]: added function comment

File: pkg/transport/transport.go [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+// copying and modifying the request
+type transport struct {
+	Modifiers []RequestModifier
+	Base      http.RoundTripper
+
+	mu     sync.Mutex                      // guards modReq
+	modReq map[*http.Request]*http.Request // original -> modified
+}
+
+func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (tiborvass) [inline pkg/transport/transport.go]: removed non-generic function comment

File: pkg/transport/transport.go [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+	if mod == nil {
+		delete(t.modReq, orig)
+	} else {
+		t.modReq[orig] = mod
+	}
+}
+
+// CloneRequest returns a clone of the provided *http.Request.
+// The clone is a shallow copy of the struct and its Header map.
+func CloneRequest(r *http.Request) *http.Request {
```
Comment (tiborvass) [inline pkg/transport/transport.go]: because authTransport needs to access the response (to cache the token), it had to implement http.RoundTripper itself and thus needed cloneRequest (which is recommended by courtesy to the caller)

File: pkg/transport/transport.go [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+	// deep copy of the Header
+	r2.Header = make(http.Header, len(r.Header))
+	for k, s := range r.Header {
+		r2.Header[k] = append([]string(nil), s...)
+	}
+
+	return r2
+}
+
+type OnEOFReader struct {
```
Comment (tiborvass) [inline pkg/transport/transport.go]: Also reusing this for `registry.authTransport` /cc @stevvooe

File: pkg/useragent/useragent.go [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+// Convert versions to a string and append the string to the string base.
+//
+// Each VersionInfo will be converted to a string in the format of
+// "product/version", where the "product" is get from the name field, while
+// version is get from the version field. Several pieces of verson information
+// will be concatinated and separated by space.
+//
+// Example:
+// AppendVersions("base", VersionInfo{"foo", "1.0"}, VersionInfo{"bar", "2.0"})
+// results in "base foo/1.0 bar/2.0".
```
Comment (tiborvass) [inline pkg/useragent/useragent.go]: @diogomonica: added example

File: pkg/useragent/useragent_test.go [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+	if vi.isValid() {
+		t.Fatalf("Expected VersionInfo to be invalid")
+	}
+	vi = VersionInfo{"foo", ""}
+	if vi.isValid() {
+		t.Fatalf("Expected VersionInfo to be invalid")
+	}
+}
+
+func TestAppendVersions(t *testing.T) {
```
Comment (tiborvass) [inline pkg/useragent/useragent_test.go]: added test

File: registry/auth.go [hunk 25]
```diff
@@ -46,7 +46,7 @@ func (auth *RequestAuthorization) getToken() (string, error) {
 	}
 
 	client := auth.registryEndpoint.HTTPClient()
-	factory := HTTPRequestFactory(nil)
+	client.Transport = transport.NewTransport(client.Transport, DockerHeaders(nil)...)
```
Comment (tiborvass) [inline registry/auth.go]: `HTTPRequestFactory` is basically replaced by `DockerHeaders`

File: registry/auth.go [hunk 26]
```diff
@@ -111,6 +111,8 @@ func loginV1(authConfig *cliconfig.AuthConfig, registryEndpoint *Endpoint, facto
 		serverAddress = authConfig.ServerAddress
 	)
 
+	client.Transport = transport.NewTransport(client.Transport, DockerHeaders(nil)...)
```
Comment (tiborvass) [inline registry/auth.go]: `DockerHeaders` comes from https://github.com/docker/docker/pull/13265/files#diff-2b07b8a4e93be324f246494ec743cde2L35 where a call to `HTTPRequestFactory` is made and passed down to `Login()` and then to `loginV1()` (this function)

File: registry/auth.go [hunk 27]
```diff
@@ -92,21 +92,23 @@ func (auth *RequestAuthorization) Authorize(req *http.Request) error {
 }
 
 // Login tries to register/login to the registry server.
-func Login(authConfig *cliconfig.AuthConfig, registryEndpoint *Endpoint, factory *requestdecorator.RequestFactory) (string, error) {
+func Login(authConfig *cliconfig.AuthConfig, registryEndpoint *Endpoint) (string, error) {
+	client := registryEndpoint.HTTPClient()
+	client.Transport = transport.NewTransport(client.Transport, DockerHeaders(nil)...)
```
Comment (tiborvass) [inline registry/auth.go]: `DockerHeaders` replaces the `HTTPRequestFactory` from the only caller https://github.com/docker/docker/pull/13265/files#diff-2b07b8a4e93be324f246494ec743cde2L35

File: registry/endpoint.go [hunk 28]
```diff
@@ -171,12 +170,14 @@ func (e *Endpoint) pingV1(factory *requestdecorator.RequestFactory) (RegistryInf
...
 
-	req, err := factory.NewRequest("GET", e.Path("_ping"), nil)
+	req, err := http.NewRequest("GET", e.Path("_ping"), nil)
 	if err != nil {
 		return RegistryInfo{Standalone: false}, err
 	}
 
-	resp, _, err := doRequest(req, nil, ConnectTimeout, e.IsSecure)
+	client := e.HTTPClient()
+	client.Transport = transport.NewTransport(client.Transport, DockerHeaders(nil)...)
```
Comment (tiborvass) [inline registry/endpoint.go]: `DockerHeaders` replaces `HTTPRequestFactory` from https://github.com/docker/docker/pull/13265/files#diff-1122d19f2eb2acdcfa7d7093b81b18e9L138

File: registry/endpoint.go [hunk 29]
```diff
@@ -100,15 +99,18 @@ func newEndpoint(address string, secure bool) (*Endpoint, error) {
 		return nil, err
 	}
 	endpoint.IsSecure = secure
+	tr := NewTransport(ConnectTimeout, endpoint.IsSecure)
+	endpoint.client = HTTPClient(transport.NewTransport(tr, DockerHeaders(metaHeaders)...))
```
Comment (tiborvass) [inline registry/endpoint.go]: Ping @stevvooe 

Any endpoint instantiated with `newEndpoint` or `NewEndpoint` is guaranteed to have an instantiated `client` field that can then be used anywhere the endpoint is passed on.

Prior to this, I had an `HTTPClient()` method on `Endpoint` which would lazily instantiate a client without providing `DockerHeaders` which was not the behavior on master. I considered passing `metaHeaders` to the `HTTPClient()` method, but that introduced another problem, which was that the header would be ignored if the client had already be instantiated. If I made it not to ignore the header, then there was a possibility of wrapping the client's existing Transport once more.

The solution I suggest here is to ensure, that each endpoint is instantiated with `metaHeaders` and has a corresponding ready-to-use HTTP client.

File: pkg/transport/transport.go (hunk 22 above)
Comment (stevvooe) [inline pkg/transport/transport.go]: Document your struct. :japanese_ogre:

File: registry/endpoint.go (hunk 29 above)
Comment (stevvooe) [inline registry/endpoint.go]: I see. It seems like we need to get control over client instantiation. We need to differentiate between headers that are always present (Auth, ident, etc.) from those that are part of the specific request flow.

I'm not sure how much this should affect this particular PR.

File: pkg/transport/transport.go (hunk 22 above)
Comment (tiborvass) [inline pkg/transport/transport.go]: fixed

**Comment by stevvooe** [comment]
LGTM!

**Comment by dmcgowan** [comment]
LGTM

**Comment by tiborvass** [comment]
rebased (was conflicting with #13308)

File: registry/registry.go [hunk 30]
```diff
@@ -31,66 +37,38 @@ const (
...
-	httpTransport := &http.Transport{
-		DisableKeepAlives: true,
-		Proxy:             http.ProxyFromEnvironment,
-		TLSClientConfig:   &tlsConfig,
+// dockerUserAgent is the User-Agent the Docker client uses to identify itself.
+// It is populated on init(), comprising version information of different components.
+var dockerUserAgent string
+
+func init() {
+	httpVersion := make([]useragent.VersionInfo, 0, 4)
```
Comment (icecrime) [inline registry/registry.go]: The capacity seems wrong (but tbh, it's a micro optimization anyways).

Comment (tiborvass) [inline registry/registry.go]: @icecrime https://github.com/docker/docker/pull/13265/files#diff-a6df1edbda0154b42349119295812901L13

**Comment by icecrime** [comment]
LGTM
